File: two_pointers/3sum_closest.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def three_sum_closest(nums, target)
    nums.sort!
    n = nums.size
    sum_closest = min = Float::INFINITY
    (1..n - 2).each do |mid_idx|
        mid = nums[mid_idx]
        left = 0
        right = n - 1
        while left < mid_idx && mid_idx < right
            sum = nums[left] + mid + nums[right]
            return target if sum == target
            if sum > target
                if sum - target < min
                    min = sum - target
                    sum_closest = sum
                end
                right -= 1
            else
                if target - sum < min
                    min = target - sum
                    sum_closest = sum
                end
                left += 1
            end
        end
    end
    sum_closest
end

File: two_pointers/3sum_closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        n = len(nums)
        sum_closest = min_diff = float('inf')

        for mid_idx in range(1, n - 1):
            mid = nums[mid_idx]
            left = 0
            right = n - 1

            while left < mid_idx and mid_idx < right:
                sum_val = nums[left] + mid + nums[right]

                if sum_val == target:
                    return target

                if sum_val > target:
                    if sum_val - target < min_diff:
                        min_diff = sum_val - target
                        sum_closest = sum_val
                    right -= 1
                else:
                    if target - sum_val < min_diff:
                        min_diff = target - sum_val
                        sum_closest = sum_val
                    left += 1

        return sum_closest